README.md
# webSHOP
Shopping cart application integrated to webERP


Multi-level dynamic css menu structure based on sales categories - sales categories can have tranlations so customers in other languages can see the translation in their language. Also a vertical indented menu structure to view sales categories and a search function to find products.

Coded in the same accessible style as webERP - a truly integrated shopping cart solution - that creates native webERP customers and user logins, orders and receipts.
jQuery validation and user interface
The long description shows as a tool tip when the user hovers above the item. Thumbnail picture shows on search and a large image shows as a tool tip when the user hovers over the thumbnail - thus saving screen real-estate and allowing fast review of items to purchase without recourse to multiple screens.

Sales in the currency of the customer
When a customer registers they can select their currency and their preferred language - prices are re-calculated at that time. The currencies in webERP can be flagged as available to web-shop customers . However, webERP prices for the currency under review for the sales type of the default web-customer need to be defined otherwise items will not display.

Only items with prices setup in the currency and price list of the customer will display

When a customer registers in the web-shop, they are set up with a customer account - the account reference is based on a numercial sequence maintained in webERP (systype =500). Newly created customer accounts inherit all the attributes of the default web-shop customer. The default web-shop customer is defined in the web-shop configuration in webERP. The new customer accounts will have the same pricing (sales type), same terms, same area, sales rep, tax group etc. However, they can select an alternative currency (and the pricing for that currency will be used in pricing web-orders) and they can also select their own interface language. If alternative terms are agreed with the customer then these can be changed inside webERP.

Customers with a credit account are not presented with paypal or credit card payment methods - the shop only places the order for sale on credit to their account.

Payment options by:
- bank transfer
- paypal
- credit card

Variable surcharge can be applied depending on the payment method selected by the customer.
If the customer elects to pay by bank transfer, then this is a low cost alternative - saving the bank fee which can be up to 4% on credit cards/paypal. The convenience and immediacy of payment by paypal or credit card is still available and some shops may elect to have no surcharges on any payment method.

Using paypal requires a paypal account. The credit card processing uses the paypal "payments pro" API and there are fees associated with its use currently USD $30 per month plus 3.9% and $0.30 USD per transaction (not cheap!). A paypal mercant account subscribed to "Payments Pro" is required to use this credit card processing option. This is only available in the US and Canada.
Also the Payflow Pro credit card processing API is implemented which allows payments over the Payflow pro API - this is a PayPal initiative with partners available at many countries throughout the world - again a Pay Pal merchant account is required subscribed to Pay Flow pro through a PayPal partner gateway.
SwipeHQ - credit card processing also implemented - available in New Zealand only. 

All the necessary data for the store is entered via webERP - including:

- item pictures
- customer pricing
- descriptions
- long descriptions

Stock quantities on the shop front are reported from the shop inventory location only and reported as 20+ if more than 20 of an item is held. If no stock on hand in the location it is reported as "Arriving Soon".

Customer registration creates a webERP customer login for the customer account - new accounts are created with numerical account codes based on the webERP systype counter

webERP now has some extended configuration to allow the store configuration from the webERP setup module - this allows entry of html that will display on the information pages - the standard webERP includes the shop configuration page and is populated with sample English "Privacy Statement" and "Terms and Condition" which need to be reviewed as appropriate in a live installation:
- "About Us"
- "Privacy Statement"
- "Terms and Conditions"
- "Contact Us"


Item descriptions can now be held in several languages - the languages that are to be used for item descriptions can be specified in the webERP configuration now and new fields in the stock maintenance screen come up for the translations required. webERP invoices and statements now use the language of the customer.

How To setup your web shop

All the set up work is really done inside webERP.

To define the items which shold display in your web-shop, you must set up sales categories and define the items under each category. If an item is to be listed on the opening screen as a "featured" item it must be flagged as featured in webERP sales categories set up.

Also, only items that have a price defined in the currency and sales type of the default shop customer will show - if there is no price set up then the item will not even be listed.

Sales Categories are defined from the webERP main menu - Inventory->Maintenance->Sales Category Maintenance

Sales Categories are much more flexible than stock categories. It is possible to set up a heirarchy of sales categories for example in the demo we have a sales category for DVDs and underneath that we have sales categories for Drama, Action, Thriller etc then under Action we have Cruise, Shwartzeneger, Gibson - any number of levels of categories can be set up. This category heirarchy defines the menu structure for the web-shop again illustrated in the shop demo.

It doesn't matter what the stock category of the item is in webERP the sales category is completely separate and any item can be allocated to any sales category.
Only items defined as belonging to a sales category will be displayed in the web-shop so items that are not for resale to the public should not be set up under the sales category structure.

The web-shop requires configuration of the following:

The web-shop default customer and customer branch

The web-shop customer defines all the defaults for new customers set up through the shop - the currency of transactions (although the web-user can alter this), sales area, sales type (price list), sales person, tax group etc - all new customers defined through the web-shop will inherit all the settings of this default customer. All web-shop customers created will appear in webERP - using a numerical code - and if necessary any of their setting can be modified to allow then to use a special price list (sales type) or have credit terms. The default customer would normally have cash terms and if this is the case the options for payment will come up in the web-shop. If the customer has credit terms then the order will not bring up any payment options and it is assumed that the sale will be on credit.

When a customer first goes into the web-shop to order, the shop shows using all the settings of the default customer. Only when the user logs in do the settings applicable to their customer account show

Any existing webERP customer can be set up as a web-shop customer simply by adding a user login for the customer account having selected the customer there is an option to add a customer login account. However, when a customer logs into the web-shop they must use their email address as their login name not their username. When the user/customer logs in the setup of the customer will define the currency of the order, the tax treatment, credit terms etc.


doc/Change_log.txt
Change Log

28/4/18 Lisence change to GPL v 2 as per the rest of webERP
23/4/14 Added new includes/config.php options for ShopName, ShopDebtorNo and ShopBranchCode these settings in config.php over-ride the webERP shop config settings. This allows multiple webSHOP installations on the same webERP installation. However, the About Us, Terms and Conditions, FreightPolicy ContactUs pages are still configured in webERP and will be the same accross all webSHOPs off the same webERP installation 
23/4/14 ItemDetails.php was not calculating tax correctly - fixed
19/4/14 Checkout.php now processes SWIPE HQ correctly was resetting for new order before order paid
19/4/14 Payment surcharge added to order correctly was missing discount percent of 0 to insert SQL
19/4/14 Checkout.php now formats messages correctly using <div> rather than tables
29/12/13 Updates to allow AusPost freight to live with no freight
29/12/13 Add text to paypal process button to show that the button must be clicked to process payment
29/12/13 Remove place order button when paypal being displayed.
25/9/13 Ricard reported that discount was being applied twice - webERP recalculates discount matrix discount rate but webSHOP was sending net price with the order and then the discount matrix rate was recalcuated applying the discount twice. Now webSHOP creates the order with the gross price and the discount percent so the webSHOP order is the same when it displays in webERP.


includes/header.php
<?php
/* $Id: header.php 5785 2012-12-29 04:47:42Z daintree $ */

// Titles and screen header
// Needs the file config.php loaded where the variables are defined for
//  $RootPath
//  $Title - should be defined in the page this file is included with

if (!headers_sent()){
	if ($StrictXHTML) {
		header('Content-type: application/xhtml+xml; charset=utf-8');
	} else {
		header('Content-type: text/html; charset=utf-8');
	}
}
echo '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
		"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
		<html xmlns="http://www.w3.org/1999/xhtml"><head><title>' . $Title . '</title>
		<link rel="shortcut icon" href="' . $PathPrefix . 'favicon.ico" />
		<link rel="icon" href="' . $PathPrefix . 'favicon.ico" />';
if ($StrictXHTML) {
	echo '<meta http-equiv="Content-Type" content="application/xhtml+xml; charset=utf-8" />';
} else {
	echo '<meta http-equiv="Content-Type" content="application/html; charset=utf-8" />';
}

echo'	<link href="' . $RootPath . '/css/'. $Theme .'.css" rel="stylesheet" type="text/css" />
	<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
	<script type="text/javascript" src ="' . $RootPath . '/javascripts/jquery.validate.min.js"></script>
	<script type="text/javascript" src ="' . $PathPrefix . '/javascripts/MiscFunctions.js"></script>
	';

/* Google Analytics tracking code only if not in test, as test distorts statistics*/
if($_SESSION['ShopMode']!='test'){
	echo "<script type='text/javascript'>

	  var _gaq = _gaq || [];
	  _gaq.push(['_setAccount', '". $GoogleAnalyticsID ."']);
	  _gaq.push(['_trackPageview']);

	  (function() {
		var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
		ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
		var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
	  })();
	</script>";
}

echo'</head>
<body>

<div id="header_block">
	<div id="shop_logo">
		<img src="' . $RootPath . '/css/webshop_logo.jpg">
	</div>
	<div id="shop_title">
		<h1>&nbsp;' . $_SESSION['ShopName'] . '</h1>
	</div>';


if (isset($_POST['CurrCode'])){
	$CurrCode = $_POST['CurrCode'];
}elseif (isset($_GET['CurrCode'])){
	$CurrCode = $_GET['CurrCode'];
}
if (isset($CurrCode) AND $CurrCode!=$_SESSION['CustomerDetails']['currcode']){
	$_SESSION['CustomerDetails']['currcode'] = $CurrCode;
	// change as well the decimal places to show
	$SQL = "SELECT decimalplaces FROM currencies WHERE currabrev ='" . $CurrCode . "'";
	$result = DB_query($SQL,$db);
	if (DB_num_rows($result)==1){
		$NewCurrencyRow = DB_fetch_array($result);
		$_SESSION['CustomerDetails']['currdecimalplaces'] = $NewCurrencyRow['decimalplaces'];
	}
	update_currency_prices($CurrCode);
}

$Errors= array();
$CarriageReturnOrLineFeed = array("\r\n",'\r\n',"\r","\n");

if (isset($_POST['Login']) AND (!isset($_SESSION['LoggedIn']) OR $_SESSION['LoggedIn']==false)){
	//This could be run either from the header form or from the Checkout form - the checkout form now has two possible login forms
	include('includes/Login.php');
}

if (isset($_POST['AddToCart'])){ //as it could be from the ItemDetails.php page
	$_GET['AddToCart'] = $_POST['StockID'];
}

/*Process any cart additions */
if (isset($_GET['AddToCart'])){

	$_GET['Page'] = 'ShoppingCart'; //display the shopping cart with the new item in it
	if (isset($_SESSION['ShoppingCart'][$_GET['AddToCart']])){
		if (isset($_POST['Quantity'])){  //as it could be from the ItemDetails.php script
			$_SESSION['ShoppingCart'][$_GET['AddToCart']]->Quantity += filter_number_format($_POST['Quantity']);
		} else {
			//Just increment the quantity if the item is already in the shopping cart
			$_SESSION['ShoppingCart'][$_GET['AddToCart']]->Quantity++;
		}
	} else {
		$SQL = "SELECT stockid,
						description,
						longdescription,
						taxcatid,
						discountcategory,
						decimalplaces,
						mbflag,
						grossweight,
						volume
				FROM stockmaster
				WHERE stockmaster.stockid ='" . $_GET['AddToCart'] . "'";
		$result = DB_query($SQL,$db);
		if (DB_num_rows($result)==1){
			$NewItemRow = DB_fetch_array($result);
			$Description = $NewItemRow['description'];
			//need to get description translation (if any) and price grossed up for tax
			$TranslatedDescription = get_item_description_translation($_GET['AddToCart']);
			if ($TranslatedDescription!=false) {
				$Description = $TranslatedDescription;
			}
			$Price = GetPrice($_GET['AddToCart'],
								$_SESSION['ShopDebtorNo'],
								$_SESSION['ShopBranchCode'],
								$_SESSION['CustomerDetails']['currcode']);

			$Discount = GetDiscount($NewItemRow['discountcategory'],1, $db);
			$Price = $Price * (1- $Discount);
			if (isset($_POST['Quantity'])){  //as it could be from the ItemDetails.php script
				$Quantity = filter_number_format($_POST['Quantity']);
			} else {
				$Quantity =1;
			}
			$_SESSION['ShoppingCart'][$_GET['AddToCart']] = new CartItem($_GET['AddToCart'],
																			$Description,
																			$NewItemRow['longdescription'],
																			$Price,
																			$Quantity,
																			$NewItemRow['decimalplaces'],
																			$NewItemRow['taxcatid'],
																			$NewItemRow['discountcategory'],
																			$NewItemRow['grossweight'],
																			$NewItemRow['volume'],
																			$NewItemRow['mbflag'],
																			$Discount);

		} // no rows returned from query to get the item to add details
	} // its a new item not already in the shopping cart
} // end of adding an item to the shopping cart

if (isset($_GET['Delete'])) {
	unset($_SESSION['ShoppingCart'][$_GET['Delete']]);
	$_GET['Page']='ShoppingCart';
}
include('includes/RecalculateCartTotals.php');

echo '<div id="cart_summary" title="' . _('Click to show the detail of what is in the shopping cart') . '">
		<form id="CartSummaryForm" method="post" action="' . $RootPath . '/index.php">
		<input type="hidden" name="FormID" value="' . $_SESSION['FormID'] . '" />
		<table class="cart_summary">
				<tr>
					<th colspan="3">' . _('Cart Summary') . '</th>
				</tr>
				<tr>
					<td>' . ('Item count') . ':&nbsp;&nbsp;' . locale_number_format($CountItems,0) . '</td>
					<td>' . ('Total Cost') . '&nbsp;';
if (!isset($_SESSION['LoggedIn']) OR $_SESSION['LoggedIn']==false){ //if not logged in then allow the user to change her currency
	echo '<select id="Currency" class="cart_summary_currency" name="CurrCode">';
	$CurrenciesResult = DB_query("SELECT currabrev FROM currencies WHERE webcart=1",$db);
	while ($CurrRow = DB_fetch_array($CurrenciesResult)){
		echo '<option ';
		if ($_SESSION['CustomerDetails']['currcode']==$CurrRow['currabrev']){
			echo 'selected="selected" ';
		}
		echo 'value="' . $CurrRow['currabrev'] . '">' . $CurrRow['currabrev'] . '</option>';
	}
	echo '</select>';
} else { // if logged in then stuck with the currency of the account
	echo $_SESSION['CustomerDetails']['currcode'];
}

echo '</td>
	<td class="number">' . locale_number_format($TotalCartValue,$_SESSION['CustomerDetails']['currdecimalplaces']) . '</td>
	</tr>
	</form>
	</table>
	</div>
			</div>';//end header_block
echo '<div class="navbar">
		<div class="top-links">
			<ul>
			<li><a href="index.php">' . _('Home') . '</a></li>
			<li><a href="index.php?Page=ShoppingCart">' . _('View Order') . '</a></li>
			<li><a href="Checkout.php">' . _('Checkout') . '</a></li>';

if (isset($_SESSION['LoggedIn']) AND $_SESSION['LoggedIn']==true){ // need to get the user details for modification
	echo '<li class="header_login">' . _('Logged in as') . ' ' . $_SESSION['UsersRealName'] . '</li>
		<li><a href="Register.php">' . _('Update Account Details') . '</a></li>
		<li><a href="index.php?LoggOff=1" onclick="return confirm(\''._('Are you sure you wish to logout?').'\');">' . _('Logout') . '</a>';
} else {
	echo '<li><a href="Register.php">' . _('Register') . '</a></li>
		<li class="header_login"><form id="LoginForm" method="post" action="' . htmlspecialchars($_SERVER['PHP_SELF'],ENT_QUOTES,'UTF-8') . '">
			<input type="hidden" name="FormID"
				value="' . $_SESSION['FormID'] . '" />
			<input type="email" placeholder="' . _('Enter Email') . '" class="required username ' . (in_array('UserEmail',$Errors) ?  'error' : '' ) . '" required="required" name="UserEmail" size="20" maxlength="30" value="" />
			<input type="password" placeholder="' . _('Password') . '" required="required" class="required password ' . (in_array('Password',$Errors) ?  'error' : '' ) . '" name="Password" size="15" maxlength="15" />
			<input class="button" type="submit" name="Login" value="' . _('Login') . '" />
		</form>';
}
echo '			</li>
			</ul>
		</div>';
echo '<div id="search_description">
		<form id="SearchForm" method="post" action="' . $RootPath . '/index.php">
		<input type="hidden" name="FormID" value="' . $_SESSION['FormID'] . '" />

	<img src="css/button-search.png">
	<input type="search" placeholder="' . _('Search') . '" name="SearchDescription" value="" onclick="this.value = \'\';" onkeydown="this.style.color = \'#000000\';" />
	</form>
	</div>
</div>';//end header_block

?>
